src/main.rs

```diff
@@ -274,57 +289,84 @@ impl BFInterp {
                 Type::IntegerType { bits } => match bits {
                     8 => todo!(),
                     16 => todo!(),
                     32 => {
-                        let op0 = self.get_i32_op(self.vars.get(name).unwrap()); //int_vars
-                        let op1 = self.get_i32_op(op1);
-                        self.i32_bin_operation(op0, op1, dest, operation_type)
+                        let op0 = self.get_int_op::<i32>(self.vars.get(name).unwrap()); //int_vars
+                        let op1 = self.get_int_op(op1);
+                        self.int_bin_operation(op0, op1, dest, operation_type)
                     }
                     64 => {
-                        let op0 = self.get_i64_op(self.vars.get(name).unwrap());
-                        let op1 = self.get_i64_op(op1);
-                        self.i64_bin_operation(op0, op1, dest, operation_type)
+                        let op0 = self.get_int_op::<i64>(self.vars.get(name).unwrap());
+                        let op1 = self.get_int_op(op1);
+                        self.int_bin_operation(op0, op1, dest, operation_type)
                     }
                     128 => todo!(),
                     _ => todo!(),
                 },
                 _ => todo!(),
             },
             ConstantOperand(con_op) => match con_op.as_ref() {
-                Constant::Int { value, .. } => {
-                    *value as i32;
-                }
-                Constant::Vector(_) => todo!(),
-                _ => todo!(),
-            },
-            MetadataOperand => todo!(),
-        }
-    }
-
-    fn get_i32_op(&self, op: &Operand) -> i32 {
-        match op {
-            LocalOperand { name, .. } => self.get_i32_op(self.vars.get(name).unwrap()),
-            ConstantOperand(con_op) => match con_op.as_ref() {
-                Constant::Int { value, .. } => *value as i32,
+                Constant::Int { .. } => todo!(),
                 Constant::Vector(_) => todo!(),
-                Constant::Undef(..) => panic!(),
                 _ => todo!(),
             },
             MetadataOperand => todo!(),
         }
     }
 
-    fn get_i64_op(&self, op: &Operand) -> i64 {
+    fn get_int_op<
+        T: 'static
+            + std::convert::TryFrom<i8>
+            + std::convert::TryFrom<i16>
+            + std::convert::TryFrom<i32>
+            + std::convert::TryFrom<i64>
+            + std::convert::TryFrom<i128>,
+    >(
+        &self,
+        op: &Operand,
+    ) -> T {
         match op {
-            LocalOperand { name, .. } => self.get_i64_op(self.vars.get(name).unwrap()),
+            LocalOperand { name, .. } => self.get_int_op(self.vars.get(name).unwrap()),
             ConstantOperand(con_op) => match con_op.as_ref() {
-                Constant::Int { value, .. } => *value as i64,
+                Constant::Int { value, .. } => {
+                    if TypeId::of::<T>() == TypeId::of::<i8>() {
```

Comment:
I see what you mean now about this being more complicated than I thought. I need to have a think about this. It might be that macros will work better in this case.

Reply:
Hmm, I don't quite understand why we need to do the `TypeId::of` dance here. We know that `value` is always a `u64`. Why can't we do something like this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a34cdde76ac50a4499f4c78442113fc0

Reply:
Also, remind me why we are casting to `i8` here and not `u8`.

Reply:
So if I remember correctly try_from goes to Err() when doing at least some conversions whereas using 'as' seems to work. So I want to do the conversion using as. However I don't think as_() works for casting to T so I need to cast to a specific type like i8, i16 etc for which reason we need the TypeId stuff.

In terms of types, the only information we have is the number of bits e.g. 8 or 16 but not whether the integer is signed or unsigned. I should do some tests to see if everything works using signed types for unsigned types.

Reply:
I don't think I quite follow. The example in rust playground compiles. You can play around in there to see if it works with more types. I might be wrong, but I think you can just do
```
                        match T::try_from(value) {
                            Ok(val) => return val,
                            Err(_) => panic!(),
                        }
```
since we know `value` is always `u64`. At the moment it looks like you are doing two casts, which appear to be the same: first you do `let val: i8 = value.as_()` and then you to `match i8::try_from(val)`. One of those casts seems unnecessary. Though there might be a reason I'm missing.

Reply:
Ok I've put an example in the playground.

Reply:
Yes, but are we expected this to happen? Presumably, we have already checked with the `bits` flag, that the value we are reading out fits. For example, if bits is `32`, we assume that the value fits into `u32` (`i32`?), so that calling `get::<u32>` can never reach the `unreachable` code.

Reply:
So that number works with get::<u32> and it gets 4294967293 but not with get::<i32>. When using 'as' that number gets converted to -3 which is the number wanted, not 4294967293.

Reply:
Aha! I see the problem now. So we either use macros, or, even simpler, you make `get_int_op` always return a `u64` and then in `int_bin_operations` you just do `let op0 = self.get_int_op() as i8/i16/i32;` (depending on the `bits`).

Reply:
I should have seen that. Thanks.